rust/cadeau/src/xmf/vpx/mod.rs

```diff
@@ -0,0 +1,225 @@
+
+use xmf_sys::{
+    vpx::{XmfVpxCodecType, XmfVpxPacketKind},
+    XmfVpxFrame, XmfVpxFrame_Destroy, XmfVpxFrame_GetBuffer, XmfVpxFrame_GetDuration, XmfVpxFrame_GetFlags,
+    XmfVpxFrame_GetHeight, XmfVpxFrame_GetPartitionId, XmfVpxFrame_GetPts, XmfVpxFrame_GetSize,
+    XmfVpxFrame_GetSpatialLayerEncoded, XmfVpxFrame_GetWidth, XmfVpxImage, XmfVpxImage_Destroy, XmfVpxPacket,
+    XmfVpxPacket_Destroy, XmfVpxPacket_GetFrame, XmfVpxPacket_GetKind, XmfVpxPacket_IsEmpty,
+};
+
+pub mod decoder;
+pub mod encoder;
+
+#[derive(Debug, Clone, Copy)]
+pub enum VpxCodec {
+    VP8,
+    VP9,
+}
+
+pub struct VpxImage {
+    pub(crate) ptr: *const XmfVpxImage,
+}
+
+impl Drop for VpxImage {
+    fn drop(&mut self) {
+        // Safety: it is safe to call, the owenership of the pointer is managed by the XmfVpxImage itself.
+        unsafe {
+            XmfVpxImage_Destroy(self.ptr.cast_mut());
+        }
+    }
+}
+
+impl From<VpxCodec> for XmfVpxCodecType {
+    fn from(codec: VpxCodec) -> Self {
+        match codec {
+            VpxCodec::VP8 => XmfVpxCodecType::VP8,
+            VpxCodec::VP9 => XmfVpxCodecType::VP9,
+        }
+    }
+}
+
+pub fn is_key_frame(buffer: &[u8]) -> bool {
+    if buffer.is_empty() {
+        return false;
+    }
+    buffer[0] & 0x1 == 0
+}
+
+pub struct VpxPacket {
+    pub(crate) ptr: *const XmfVpxPacket,
+}
+
+impl VpxPacket {
+    /// # Safety
+    ///
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn kind(&self) -> XmfVpxPacketKind {
+        // Safety: It's up to the caller to ensure the packet is valid.
+        unsafe { XmfVpxPacket_GetKind(self.ptr.cast_mut()) }
+    }
+
+    /// # Safety
+    ///
+    /// Obtain the deep copy of the frame from the packet, must be used before any call to VpxEncoder method.
+    /// The frame produced by this method is always valid.
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn frame(&self) -> Option<VpxFrame> {
+        // Safety: it's up to the caller to ensure the packet is valid, the frame will remain valid until the frame itself is destroyed.
+        let frame_ptr = unsafe { XmfVpxPacket_GetFrame(self.ptr) };
+        if frame_ptr.is_null() {
+            None
+        } else {
+            VpxFrame::new(frame_ptr).ok()
+        }
+    }
+
+    /// # Safety
+    ///
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn is_empty(&self) -> bool {
+        // Safety: It's up to the caller to ensure the packet is valid.
+        unsafe { XmfVpxPacket_IsEmpty(self.ptr) }
+    }
```

Comment:
issue: The `cadeau` crate is supposed to be a easy-to-use wrapper on top of the `libxmf`. Avoid exposing unsafe functions by checking the required invariants yourself and expose safe wrappers. In this case, the pointer being valid should be an invariant of `VpxPacket`, allowing you to call these functions safely without unsafe block. The problem here is that you are eliding the lifetime, allowing the underlying structure to be freed while we are still borrowing it. If we are borrowing, introduce a lifetime using `PhantomData`.

```rust
pub struct VpxPacket<'a> {
    ptr: *const XmfVpxPacket,
    _phantom: PhantomData<&'a XmfVpxPacket>,
}
```

Conceptually, this is telling that the structure should be checked as-if it contained a reference to a XmfVpxPacket.

Also, the `pub(crate)` in `pub(crate) ptr: *const XmfVpxPacket` is a a troublesome pattern. Conceptually this field is unsafe because there is safety invariants associated to it, but there is currently no way of specifying this. As such, keep it private and expose a `pub(crate) unsafe from_raw()`  function or similar, with proper safety invariants documentation.

Also that being said, if we are indeed borrowing, drop should definitely not be called. You could define a owned version taking a `ptr: *mut XmfVpxPacket`, or turn this into an enum to make runtime checks.

```rust
pub struct VpxPacket<'a>(VpxPacketInner<'a>);

enum VpxPacketInner<'a> {
    Borrowed { ptr: *const XmfVpxPacket, _phantom: PhantomData<'a XmfVpxPacket> },
    Owned { ptr: *mut XmfVpxPacket },
}
```